File: 20240411_Thursday/20240411.py
# https://leetcode.com/problems/remove-k-digits/discuss/1779520/Python3-MONOTONIC-STACK-(oo)-Explained
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        st = list()
        for n in num:
            while st and k and st[-1] > n:
                st.pop()
                k -= 1
            
            if st or n is not '0': # prevent leading zeros
                st.append(n)
                
        if k: # not fully spent
			st = st[0:-k]
            
        return ''.join(st) or '0'

File: 20240411_Thursday/20240411.go
// Leetcode: 402. Remove K Digits.
// https://leetcode.com/problems/remove-k-digits/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 3 ms, faster than 69.77% of Go online submissions for Remove K
// Digits.
// Memory Usage: 4.4 MB, less than 96.51% of Go online submissions for Remove
// K Digits.
// 2024.04.11 Daily Challenge.

package main

func removeKdigits(num string, k int) string {
	stk := make([]byte, len(num))
	i := -1
	for _, v := range num {
		v := byte(v)
		for i >= 0 && k > 0 && stk[i] > v {
			i -= 1
			k -= 1
		}
		if i >= 0 || '0' != v {
			i += 1
			stk[i] = v
		}
	}
	if -1 == i {
		return "0"
	}
	if k >= i+1 {
		return "0"
	}
	if k > 0 {
		stk = stk[0 : i+1-k]
	} else {
		stk = stk[0 : i+1]
	}
	if 0 == len(stk) {
		return "0"
	}
	return string(stk)
}
